setuptools/dist.py

```diff
@@ -138,11 +139,15 @@ def invalid_unless_false(dist, attr, value):
     raise DistutilsSetupError(f"{attr} is invalid.")
 
 
-def check_requirements(dist, attr, value):
+@overload
+def check_requirements(dist, attr: str, value: set) -> NoReturn: ...
+@overload
+def check_requirements(dist, attr: str, value: _StrOrIter) -> None: ...
```

Comment:
I think `_StrOrIter` is wrong here. An `Iterable` could be an `Iterator` that can be exhausted after the first `_reqs.parse()`. What we really want here is an `OrderedSequence`, which might be defined as every `Sequence` that's not unordered. I'm not sure if it's possible to define that in mypy, so an approximation is fine (maybe even `_StrOrIter`).